File: telecomcustomerserviceanalysis/telecom-customer/src/main/java/com/edu/bigdata/consumer/util/ScanRowkeyUtil.java
package com.edu.bigdata.consumer.util;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.List;

/**
 * 该类主要用于根据用户传入的手机号以及开始和结束时间点，按月生成多组 rowKey
 */
public class ScanRowkeyUtil {

    private String telephone;
    private String startDateString;
    private String stopDateString;
    List<String[]> list = null;

    int index = 0;

    private SimpleDateFormat sdf1 = new SimpleDateFormat("yyyy-MM-dd");
    private SimpleDateFormat sdf2 = new SimpleDateFormat("yyyyMMddHHmmss");

    public ScanRowkeyUtil(String telephone, String startDateString, String stopDateString) {
        this.telephone = telephone;
        this.startDateString = startDateString;
        this.stopDateString = stopDateString;

        list = new ArrayList<>();
        genRowKeys();
    }

    // 01_15837312345_201711
    // 15837312345 2017-01-01 2017-05-01
    public void genRowKeys() {
        int regions = Integer.valueOf(PropertiesUtil.getProperty("hbase.calllog.regions.count"));
        try {
            Date startDate = sdf1.parse(startDateString);
            Date stopDate = sdf1.parse(stopDateString);

            // b当前开始时间
            Calendar currentStartCalendar = Calendar.getInstance();
            currentStartCalendar.setTimeInMillis(startDate.getTime());
            // 当前结束时间
            Calendar currentStopCalendar = Calendar.getInstance();
            currentStopCalendar.setTimeInMillis(startDate.getTime());
            currentStopCalendar.add(Calendar.MONTH, 1);

            while (currentStopCalendar.getTimeInMillis() <= stopDate.getTime()) {

                String regionCode = HBaseUtil.genRegionCode(telephone, sdf2.format(new Date(currentStartCalendar.getTimeInMillis())), regions);
                // 01_15837312345_201711
                String startRowKey = regionCode + "_" + telephone + "_" + sdf2.format(new Date(currentStartCalendar.getTimeInMillis()));
                String stopRowKey = regionCode + "_" + telephone + "_" + sdf2.format(new Date(currentStopCalendar.getTimeInMillis()));

                String[] rowkeys = {startRowKey, stopRowKey};
                list.add(rowkeys);
                currentStartCalendar.add(Calendar.MONTH, 1);
                currentStopCalendar.add(Calendar.MONTH, 1);
            }
        } catch (ParseException e) {
            e.printStackTrace();
        }
    }

    /**
     * 判断 list 集合中是否还有下一组 rowKey
     *
     * @return
     */
    public boolean hasNext() {

        if (index < list.size()) {
            return true;
        } else {
            return false;
        }
    }

    /**
     * 取出 list 集合中存放的下一组 rowKey
     *
     * @return
     */
    public String[] next() {

        String[] rowkeys = list.get(index);
        index++;
        return rowkeys;
    }
}

File: telecomcustomeranalysis/telecom-customer/src/main/scala/com/edu/bigdata/customer/util/ScanRowkeyUtil.scala
package com.edu.bigdata.customer.util

import java.util.{Calendar, Date}

import org.apache.commons.lang3.time.FastDateFormat

import scala.collection.mutable.ListBuffer

class ScanRowkeyUtil extends Serializable {

  var telephone:String = _
  var startDateString:String = _
  var stopDateString:String = _
  var list:ListBuffer[Array[String]] = _
  var index:Int = 0

  val df1 = new FastDateFormat("yyyy-MM-dd")
  val df2 = new FastDateFormat("yyyyMMddHHmmss")


  this(telephone:String , startDateString:String , stopDateString:String ) {
    this.telephone = telephone;
    this.startDateString = startDateString;
    this.stopDateString = stopDateString;

    list = new ListBuffer[Array[String]]()
    genRowKeys()
  }

  // 01_15837312345_201711
  // 15837312345 2017-01-01 2017-05-01
  def genRowKeys():Unit={
    val regions:Int = PropertiesUtil.getProperty("hbase.calllog.regions.count").toInt
    val startDate = df1.parse(startDateString)
    val stopDate = df2.parse(stopDateString)
    val currentStartCalendar = Calendar.getInstance()
    currentStartCalendar.setTimeInMillis(startDate.getTime())
    val currentStopCalendar = Calendar.getInstance()
    currentStopCalendar.setTimeInMillis(stopDate.getTime())
    currentStopCalendar.add(Calendar.MONTH,1)

    while (stopDate.getTime() >= currentStopCalendar.getTimeInMillis){
      val regionCode = HBaseUtil.genRegionCode(telephone,df2.format(new Date(currentStartCalendar.getTimeInMillis())),regions)
      val startRowKey:String = regionCode + "_" + telephone + "_" + df2.format(new Date(currentStartCalendar.getTimeInMillis()));
      val stopRowKey:String = regionCode + "_" + telephone + "_" + df2.format(new Date(currentStopCalendar.getTimeInMillis()));

      val rowkeys:Array[String] = Array(startRowKey, stopRowKey);
      list.append(rowkeys);
      currentStartCalendar.add(Calendar.MONTH, 1);
      currentStopCalendar.add(Calendar.MONTH, 1);
    }

  }

  /**
    * 判断 list 集合中是否还有下一组 rowKey
    *
    * @return
    */

  def hasNext(): Boolean ={
    if (index < list.size){
      true
    }else{
      false
    }
  }

  /**
    * 取出 list 集合中存放的下一组 rowKey
    *
    * @return
    */

  def next():Array[String]={
    val rowkeys = list(index)
    index = index + 1
    rowkeys
  }

}
